Added function that converts TSE to phyloseq

TuomasBorman opened the pull request:
Hi!

Here is the function that creates a phyloseq object from TSE. There is one problem with tests. Even though phyloseq functions are like "phyloseq::otu_table()", they are not found. I tested run those same codes outside from "testthat" and they work just fine, and everything seems to be in right place in the phyloseq object. 

Also, because I did not found any data that contains referenceseq, I could not test that it goes to the right place in the phyloseq object.

-Tuomas

TuomasBorman committed:
Added function that converts TSE to phyloseq

antagomir commented:
Did you try to add: `library(phyloseq)` to load it in the test file?

antagomir reviewed (commented):
Seems fine. We can discuss the problem with tests separately in the PR comments.

TuomasBorman committed:
Fixed a bug

TuomasBorman commented:
> Did you try to add: `library(phyloseq)` to load it in the test file?

Ok, found the error. The error was that the function itself did not have "phyloseq::" when using phyloseq's functions

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+#' Create a phyloseq object from a TreeSummarizedExperiment object
+#'
```
I know its quite redundant, but a description is needed.

TuomasBorman replied:
Added

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+#' @examples
+#' # Get TSE object
+#' data("GlobalPatterns")
+#' TSE <- GlobalPatterns
```
TSE -> tse?

Consistency is king

TuomasBorman replied:
Yep, true

Changed

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+#' TSE <- GlobalPatterns
+#'
+#' # Create a phyloseq object from it
+#' phy <- makePhyloseqFromTreeSummarizedExperiment(TSE)
```
For generating copy & paste examples and also generating nice pkgdown output, add a line printing the `phy`object.
Like this:

```
phy <- makePhyloseqFromTreeSummarizedExperiment(TSE)
phy
```

TuomasBorman replied:
Added

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+#'
+#' # Counts relative abundances table
+#' TSE <- transformCounts(TSE, method = "relabundance")
+#' phy2 <- makePhyloseqFromTreeSummarizedExperiment(TSE, abund_values = "relabundance")
```
See above

TuomasBorman replied:
TSE --> tse
added phy2 printing

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+            # Check object
+            if(!is(x,"TreeSummarizedExperiment")){
+                stop("'x' must be a 'TreeSummarizedExperiment' object")
+            }
```
This has to be restructured. Since you know, which type of object your are handling, you can handle all the different `SE` classes in a nested way.

```
setGeneric("makePhyloseqFromTreeSummarizedExperiment", signature = c("x"),
            function(x, ...)
                standardGeneric("makePhyloseqFromTreeSummarizedExperiment"))

setMethod("makePhyloseqFromTreeSummarizedExperiment", signature = c(x = "SummarizedExperiment"),
           function(x, abund_values = "counts"){
        # do conversion to all the counts data, sample data and feature data
    }
)

setMethod("makePhyloseqFromTreeSummarizedExperiment", signature = c(x = "TreeSummarizedExperiment"),
           function(x, ...){
        ans <- callNextMethod()
        # add tree data if present, add sequence data if present
        ans
    }
)
```

TuomasBorman replied:
Restructured

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+            } else {
+                tax_table <- NULL
+                tax_table <- phyloseq::tax_table(tax_table, errorIfNULL = FALSE)
+            }
```
Does phyloseq support non-taxonomic data as tax_table? I don't think so, since it is a matrix and therefore can only be of type `character`.

It would be best to subset to taxonomic data only using `rowData(x)[,taxonomyRanks(x)]`

the else is also not necessary.

TuomasBorman replied:
Changed

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+
+            # If rowData has information, stores it to tax_table and converts is to phyloseq's tax_table.
+            # Otherwise, tax_table is NULL, and it is converted to phyloseq's tax_table.
+            if(nrow(rowData(x))>0 && ncol(rowData(x))>0){
```
`nrow(rowData(x))` is never zero. (Except, if the SE has zero rows). In this case I would just refuse to convert the SE with an initial check

TuomasBorman replied:
Added all(dim(x) > 0) at the beginning

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+
+            # If colData has information, stores it to sample_data and converts is to phyloseq's sample_data.
+            # Otherwise, sample_data is NULL, and it is converted to phyloseq's sample_data.
+            if(nrow(colData(x))>0 && ncol(colData(x))>0){
```
See above. Just check that `all(dim(x) > 0)` at the beginning and be done with it

TuomasBorman replied:
Added checking

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+            } else {
+                sample_data <- NULL
+                sample_data <- phyloseq::sample_data(sample_data, errorIfNULL = FALSE)
+            }
```
Not necessary.

TuomasBorman replied:
Removed

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+            } else {
+                phy_tree <- NULL
+                phy_tree <- phyloseq::phy_tree(phy_tree, errorIfNULL = FALSE)
+            }
```
Not necessary

TuomasBorman replied:
Removed

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+            } else {
+                refseq <- NULL
+                refseq <- phyloseq::refseq(refseq, errorIfNULL = FALSE)
+            }
```
Needs to be restructured. See above.

else code path is not necessary

TuomasBorman replied:
Restructured

FelixErnst commented on R/makephyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,113 @@
+                                 tax_table,
+                                 sample_data,
+                                 phy_tree,
+                                 refseq)
```
Since the phyloseq constructor dispatches only to `...`, it would be better to construct a list of arguments and call `do.call(phyloseq::phyloseq,args)`

With this call you don't need to set elements ,if they are not present. This removes close to 50% of the code you need to maintain in this function.

TuomasBorman replied:
Code is now using do.call()

FelixErnst commented on tests/testthat/test-5makePhyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,47 @@
+    phy <- makePhyloseqFromTreeSummarizedExperiment(GlobalPatterns)
+
+    # Test that assay is in otu_table
+    testthat::expect_equal(as.data.frame(phyloseq::otu_table(phy)@.Data), as.data.frame(assays(tse)$counts))
```
the `testthat::` prefix is not necessary. If you want to test this locally either load `testthat` in your session or use `devtools::test(filter="5makePhyloseq")`

TuomasBorman replied:
Removed testthat::

FelixErnst commented on tests/testthat/test-5makePhyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,47 @@
+context("makePhyloseqFromTreeSummarizedExperiment")
```
This is an IO function in the most general sense. Please move the testing to the IO test file `test-4IO.R`

TuomasBorman replied:
Moved

FelixErnst reviewed (changes requested):
Hi Tuomas,

thanks for the PR. The code needs to be restructured a bit. Let me know, if anything is unclear.

Felix

FelixErnst commented on tests/testthat/test-5makePhyloseqFromTreeSummarizedExperiment.R:
```diff
@@ -0,0 +1,47 @@
+
+    # Test that referenceSeq is in refseq. There is no referenceseq in TSE, so
+    # referenceseq slot is not made in the phyloseq object
+    #testthat::expect_equal(phyloseq::refseq(phy), referenceSeq(tse))
```
The referenceSeq slot is present, but the value `NULL`

TuomasBorman replied:
Now the function has new structure.

So, if tse's referenceSeq has value NULL, phyloseq's refseq should be also NULL? 

Now the function does not add NULLs, because e.g. 'phyloseq' package's 'GlobalPatterns' data does not have NULL in refseq even though the corresponding tse have NULL.

FelixErnst replied:
The comment was a comment on the comment 😄 

`TSE` has a `referenceSeq` slot, which by default is `NULL`. So, yes, if `is.null(referenceSeq(tse))`, the `phyloseq` object also shouldn't have sequence data. How `phyloseq` internally stores the sequence, I have never bothered to look up.

TuomasBorman replied:
Oh, okay, that explains...

TuomasBorman committed:
Sync

Merge branch 'master' into makephyloseqFromTreeSummarizedExperiment

# Conflicts:
#	NAMESPACE

TuomasBorman committed:
Restructured function

TuomasBorman committed:
Minor change to test

FelixErnst commented:
Can you fix the error from the builds?

TuomasBorman committed:
Added '...' parameter to roxygen

TuomasBorman committed:
Corrected spacing
